plugins/ksp/src/org/jetbrains/kotlin/ksp/symbol/impl/binary/KSClassifierReferenceDescriptorImpl.kt

```diff
@@ -5,28 +5,54 @@
+import org.jetbrains.kotlin.descriptors.ClassifierDescriptor
+import org.jetbrains.kotlin.descriptors.ClassifierDescriptorWithTypeParameters
 import org.jetbrains.kotlin.ksp.symbol.KSClassifierReference
+import org.jetbrains.kotlin.ksp.symbol.KSTypeArgument
 import org.jetbrains.kotlin.ksp.symbol.Origin
 import org.jetbrains.kotlin.types.KotlinType
+import org.jetbrains.kotlin.types.TypeProjection
 
-class KSClassifierReferenceDescriptorImpl(val kotlinType: KotlinType) : KSClassifierReference {
+class KSClassifierReferenceDescriptorImpl(val descriptor: ClassifierDescriptor, val arguments: List<TypeProjection>) :
+    KSClassifierReference {
     companion object {
-        private val cache = mutableMapOf<KotlinType, KSClassifierReferenceDescriptorImpl>()
+        private val cache = mutableMapOf<Pair<ClassifierDescriptor, List<TypeProjection>>, KSClassifierReferenceDescriptorImpl>()
 
-        fun getCached(kotlinType: KotlinType) = cache.getOrPut(kotlinType) { KSClassifierReferenceDescriptorImpl(kotlinType) }
+        fun getCached(kotlinType: KotlinType) = cache.getOrPut(
+            Pair(
+                kotlinType.constructor.declarationDescriptor!!,
+                kotlinType.arguments
+            )
+        ) { KSClassifierReferenceDescriptorImpl(kotlinType.constructor.declarationDescriptor!!, kotlinType.arguments) }
+
+        fun getCached(descriptor: ClassifierDescriptor, arguments: List<TypeProjection>) =
+            cache.getOrPut(Pair(descriptor, arguments)) { KSClassifierReferenceDescriptorImpl(descriptor, arguments) }
+    }
+
+    private val nDeclaredArgs by lazy {
+        (descriptor as? ClassifierDescriptorWithTypeParameters)?.declaredTypeParameters?.size ?: 0
     }
 
     override val origin = Origin.CLASS
 
-    override val qualifier: KSClassifierReference?
-        get() = TODO("not implemented") //To change initializer of created properties use File | Settings | File Templates.
+    override val qualifier: KSClassifierReference? by lazy {
+        val outerDescriptor = descriptor.containingDeclaration as? ClassifierDescriptor ?: return@lazy null
+        val outerArguments = arguments.drop(nDeclaredArgs)
+        getCached(outerDescriptor, outerArguments)
+    }
 
     override val typeArguments: List<KSTypeArgument> by lazy {
-        kotlinType.arguments.map { KSTypeArgumentDescriptorImpl.getCached(it) }
+        arguments.map { KSTypeArgumentDescriptorImpl.getCached(it) }
     }
 
     override fun referencedName(): String {
-        return kotlinType.constructor.declarationDescriptor?.name?.asString() ?: ""
+        val declaredArgs = if (arguments.isEmpty() || nDeclaredArgs == 0)
+            emptyList()
+        else
+            arguments.subList(0, nDeclaredArgs)
+        return descriptor.name.asString() + if (declaredArgs.isNotEmpty()) "<${declaredArgs.map { it.toString() }
+            .joinToString(", ")}>" else ""
     }
+
+    override fun toString() = referencedName()
```

Comment:
Nit: Since we are just calling referencedName() in toString(), does it make sense to remove either of them?

Reply:
Not sure if I follow. Can you give an example?

Reply:
All toString() override for ClassifierReferenceDescriptor impls are simply just calling referencedName, so maybe make sense to just shrink them.